core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateUnnest.java

```diff
@@ -0,0 +1,402 @@
+        Symbol uniqueSymbol = context.getSymbolAllocator().newSymbol("unique", BIGINT);
+        PlanNode input = new AssignUniqueId(
+                context.getIdAllocator().getNextId(),
+                correlatedJoinNode.getInput(),
+                uniqueSymbol);
+
+        // pre-project unnest symbols if they were pre-projected in subquery
+        // The correlated UnnestNode either unnests correlation symbols directly, or unnests symbols produced by a projection that uses only correlation symbols.
+        // Here, any underlying projection that was a source of the correlated UnnestNode, is appended as a source of the rewritten UnnestNode.
+        // If the projection is not necessary for UnnestNode (i.e. it does not produce any unnest symbols), it should be pruned afterwards.
+        PlanNode unnestSource = context.getLookup().resolve(unnestNode.getSource());
+        if (unnestSource instanceof ProjectNode) {
+            ProjectNode sourceProjection = (ProjectNode) unnestSource;
+            input = new ProjectNode(
+                    sourceProjection.getId(),
+                    input,
+                    Assignments.builder()
+                            .putIdentities(input.getOutputSymbols())
+                            .putAll(sourceProjection.getAssignments())
+                            .build());
+        }
+
+        // determine join type for rewritten UnnestNode
+        Type unnestJoinType = LEFT;
+        if (enforceSingleRow.isEmpty() && correlatedJoinNode.getType() == CorrelatedJoinNode.Type.INNER && unnestNode.getJoinType() == INNER) {
+            unnestJoinType = INNER;
+        }
+
+        // make sure that the rewritten node is with ordinality, which might be necessary to restore inner unnest semantics after rewrite.
+        Symbol ordinalitySymbol = unnestNode.getOrdinalitySymbol().orElseGet(() -> context.getSymbolAllocator().newSymbol("ordinality", BIGINT));
+
+        // rewrite correlated join to UnnestNode.
+        UnnestNode rewrittenUnnest = new UnnestNode(
+                context.getIdAllocator().getNextId(),
+                input,
+                input.getOutputSymbols(),
+                unnestNode.getMappings(),
+                Optional.of(ordinalitySymbol),
+                unnestJoinType,
+                Optional.empty());
+
+        // restore all projections, LimitNodes and EnforceSingleRowNode from the subquery
+        PlanNode rewrittenPlan = rewriteNodeSequence(
+                context.getLookup().resolve(correlatedJoinNode.getSubquery()),
+                input.getOutputSymbols(),
+                ordinalitySymbol,
+                uniqueSymbol,
+                rewrittenUnnest,
+                unnestNode.getId(),
+                context.getLookup(),
+                context.getIdAllocator(),
+                context.getSymbolAllocator());
+
+        // restore INNER semantics in case when the original UnnestNode was INNER and the rewritten UnnestNode is LEFT.
```

Comment:
I have hard time understanding this snippet. Will take some rest and try again :)